File: src/lib/cpmf/methods/rst/RSTAccumulator.cpp
// =============================================================================
// PMFLib - Library Supporting Potential of Mean Force Calculations
// -----------------------------------------------------------------------------
//    Copyright (C) 2008 Petr Kulhanek, kulhanek@enzim.hu
//                       Martin Petrek, petrek@chemi.muni.cz
//
//     This program is free software; you can redistribute it and/or modify
//     it under the terms of the GNU General Public License as published by
//     the Free Software Foundation; either version 2 of the License, or
//     (at your option) any later version.
//
//     This program is distributed in the hope that it will be useful,
//     but WITHOUT ANY WARRANTY; without even the implied warranty of
//     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//     GNU General Public License for more details.
//
//     You should have received a copy of the GNU General Public License along
//     with this program; if not, write to the Free Software Foundation, Inc.,
//     51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
// =============================================================================

#include <errno.h>
#include <string.h>
#include <RSTAccumulator.hpp>
#include <ErrorSystem.hpp>

//==============================================================================
//------------------------------------------------------------------------------
//==============================================================================

CRSTAccumulator::CRSTAccumulator(void)
{
    NCoords     = 0;
    TotNBins  = 0;
}

//------------------------------------------------------------------------------

CRSTAccumulator::~CRSTAccumulator(void)
{
}

//==============================================================================
//------------------------------------------------------------------------------
//==============================================================================

void CRSTAccumulator::Load(const CSmallString& name)
{
    FILE* fin = fopen(name, "r");

    if(fin == NULL) {
        CSmallString error;
        error << "unable to open file '" << name << "' (" << strerror(errno) << ")";
        RUNTIME_ERROR(error);
    }

    Load(fin);
    fclose(fin);
}

//------------------------------------------------------------------------------

void CRSTAccumulator::Load(FILE* fin)
{
    if(fin == NULL) {
        INVALID_ARGUMENT("stream is not open");
    }

// read RST accumulator header ------------------
    char rst_id[4];
    char rst_ver[3];
    int  numofcoords = 0;

// first line can contain either two or three records
    char buffer[80];
    if( fgets(buffer,80,fin) == NULL ) {
        CSmallString error;
        error << "unable to read the first line";
        RUNTIME_ERROR(error);
    }

    int nr = sscanf(buffer,"%3s %2s %d",rst_id,rst_ver,&numofcoords);
    if( nr != 3) {
        CSmallString error;
        error << "incorrect header record (nr: " << nr << "), line '" << buffer << "'";
        RUNTIME_ERROR(error);
    }

// check ID string
    rst_id[3]='\0';
    if(strcmp(rst_id,"RST") != 0) {
        CSmallString error;
        error << "'RST' magic word was not found in the first line";
        RUNTIME_ERROR(error);
    }

    rst_ver[2]='\0';
    if(strcmp(rst_ver,"V1") != 0) {
        CSmallString error;
        error << "only RST V1 version is supported (line: " << buffer << ")";
        RUNTIME_ERROR(error);
    }

    if(numofcoords <= 0) {
        CSmallString error;
        error << "number of coordinates has to be greater than zero, but " << numofcoords << " was found";
        RUNTIME_ERROR(error);
    }

    SetNumberOfCoords(numofcoords);

// read coordinate specification ----------------
    for(int i=0; i < NCoords; i++) {
        int             id = 0;
        char            type[11];
        char            name[51];
        double          min_value = 0.0;
        double          max_value = 0.0;
        int             nbins = 0;
        int             tr = 0;

        memset(type,0,11);
        memset(name,0,51);

        // read item
        tr = fscanf(fin,"%d %10s %lf %lf %d",&id,type,&min_value,&max_value,&nbins);
        if( tr != 5 ) {
            CSmallString error;
            error << "unable to read coordinate definition, id: " << i+1 << " (" << tr << " != 5)";
            RUNTIME_ERROR(error);
        }

        // some tests
        if(id != i+1) {
            CSmallString error;
            error << "coordinate id does not match, read: " << id << ", expected: " << i+1;
            RUNTIME_ERROR(error);
        }
        if(max_value <= min_value) {
            CSmallString error;
            error << "min value is not smaller than max value, id: " << id;
            RUNTIME_ERROR(error);
        }
        if(nbins <= 0) {
            CSmallString error;
            error << "number of bins has to be grater than zero, id: " << id;
            RUNTIME_ERROR(error);
        }

        // read item
        tr = fscanf(fin,"%d %55s",&id,name);
        if( tr != 2 ) {
            CSmallString error;
            error << "unable to read coordinate definition, id: " << i+1 << " (" << tr << " != 2)";
            RUNTIME_ERROR(error);
        }
        // some tests
        if(id != i+1) {
            CSmallString error;
            error << "coordinate id does not match, read: " << id << ", expected: " << i+1;
            RUNTIME_ERROR(error);
        }

        // init coordinate
        SetCoordinate(i,name,type,min_value,max_value,nbins);
    }

// alloc accumulator data -----------------------
    FinalizeAllocation();

// read accumulator data ------------------------

// samples
    for(int i=0; i < TotNBins; i++) {
        int ns = 0;
        if(fscanf(fin,"%d",&ns) != 1) {
            CSmallString error;
            error << "unable to read number of samples for bin " << i;
            RUNTIME_ERROR(error);
        }
        NSamples[i] = ns;
    }
}

//==============================================================================
//------------------------------------------------------------------------------
//==============================================================================

void CRSTAccumulator::Save(const CSmallString& name)
{
    FILE* fout = fopen(name, "w");

    if(fout == NULL) {
        CSmallString error;
        error << "unable to open file '" << name << "' (" << strerror(errno) << ")";
        RUNTIME_ERROR(error);
    }

    Save(fout);
    fclose(fout);
}

//------------------------------------------------------------------------------

void CRSTAccumulator::Save(FILE* fout)
{
    if(fout == NULL) {
        INVALID_ARGUMENT("stream is not open")
    }

    if(TotNBins == 0) {
        RUNTIME_ERROR("no data in accumulator");
    }

// 10  format(A4,1X,I3)
// 20  format(I2,1X,A5,1X,E18.11,1X,E18.11,1X,I6)
// 30  format(8(I9,1X))
// 40  format(4(E19.11,1X))

// write RST accumulator header ------------------
    if(fprintf(fout," RST %3d\n",NCoords) <= 0) {
        CSmallString error;
        error << "unable to write header";
        RUNTIME_ERROR(error);
    }

// write coordinate specification ----------------
    for(int i=0; i < NCoords; i++) {
        if(fprintf(fout,"%2d %5s %18.11e %18.11E %6d\n",i+1,(const char*)Sizes[i].Type,
                   Sizes[i].MinValue,Sizes[i].MaxValue,Sizes[i].NBins) <= 0) {
            CSmallString error;
            error << "unable to write coordinate definition id: " << i+1;
            RUNTIME_ERROR(error);
        }
    }

// write accumulator data ------------------------

// samples
    int counter;

    counter = 0;
    for(int i=0; i < TotNBins; i++) {
        if(fprintf(fout,"%9d ",NSamples[i]) <= 0) {
            CSmallString error;
            error << "unable to write number of samples for bin " << i;
            RUNTIME_ERROR(error);
        }
        if(counter % 8 == 7) fprintf(fout,"\n");
        counter++;
    }
    if(counter % 8 != 0) fprintf(fout,"\n");
}

//==============================================================================
//------------------------------------------------------------------------------
//==============================================================================

void CRSTAccumulator::SetNumberOfCoords(int numofcoords)
{
    if(numofcoords <= 0) {
        INVALID_ARGUMENT("numofcoords <= 0");
    }

    if(NCoords > 0) {
        // destroy all previous data
        Deallocate();
        Sizes.FreeVector();
        NCoords = 0;
    }

// try to allocate Sizes array
    Sizes.CreateVector(numofcoords);

// all seems to be fine - update items
    NCoords = numofcoords;
}

//------------------------------------------------------------------------------

void CRSTAccumulator::SetCoordinate(int id,
                                    const CSmallString& name,
                                    const CSmallString& type,
                                    double min_value,double max_value,int nbins)
{
    if( Sizes == NULL ){
        RUNTIME_ERROR("Sizes is NULL");
    }
    if( id < 0 || id >= NCoords ){
        INVALID_ARGUMENT("id out-of-range");
    }

    if( nbins <= 0 ){
        INVALID_ARGUMENT("nbins <= 0");
    }
    if( max_value < min_value ){
        INVALID_ARGUMENT("max_value < min_value");
    }

    if( NSamples != NULL ) {
        // it was already finalized - destroy data
        Deallocate();
    }

    Sizes[id].ID = id;
    Sizes[id].Name = name;
    Sizes[id].Type = type;
    Sizes[id].MinValue = min_value;
    Sizes[id].MaxValue = max_value;

    Sizes[id].NBins = nbins;
    Sizes[id].BinWidth = (Sizes[id].MaxValue - Sizes[id].MinValue)/Sizes[id].NBins;
    Sizes[id].Width = Sizes[id].MaxValue - Sizes[id].MinValue;
}

//==============================================================================
//------------------------------------------------------------------------------
//==============================================================================

int CRSTAccumulator::GetNumberOfCoords(void) const
{
    return(NCoords);
}

//------------------------------------------------------------------------------

int CRSTAccumulator::GetNumberOfBins(void) const
{
    return(TotNBins);
}

//------------------------------------------------------------------------------

int CRSTAccumulator::GetNumberOfBinsWithLimit(int limit) const
{
    int number = 0;
    for(int i=0; i < TotNBins; i++) {
        if(NSamples[i] >= limit) number++;
    }
    return(number);
}

//------------------------------------------------------------------------------

const CColVariable* CRSTAccumulator::GetCoordinate(int cv) const
{
    return(&Sizes[cv]);
}

//==============================================================================
//------------------------------------------------------------------------------
//==============================================================================

int CRSTAccumulator::GetNumberOfSamples(const CSimpleVector<int>& position) const
{
    int glbindex = 0;
    for(int i=0; i < NCoords; i++) {
        glbindex = glbindex*Sizes[i].GetNumberOfBins() + position[i];
    }
    return(NSamples[glbindex]);
}

//------------------------------------------------------------------------------

int CRSTAccumulator::GetGlobalIndex(const CSimpleVector<int>& position) const
{
    int glbindex = 0;
    for(int i=0; i < NCoords; i++) {
        glbindex = glbindex*Sizes[i].GetNumberOfBins() + position[i];
    }
    return(glbindex);
}

//------------------------------------------------------------------------------

int CRSTAccumulator::GetNumberOfSamples(int ibin) const
{
    return(NSamples[ibin]);
}

//==============================================================================
//------------------------------------------------------------------------------
//==============================================================================

void CRSTAccumulator::Clear(void)
{
    if(NSamples == NULL) return;   // allocation was not finalized

    for(int i=0; i < TotNBins; i++) NSamples[i] = 0;
}

//==============================================================================
//------------------------------------------------------------------------------
//==============================================================================

void CRSTAccumulator::FinalizeAllocation(void)
{
    if(NCoords <= 0) return;     // at least one coordinate is required
    if(Sizes == NULL) return;   // Sizes array is not allocated
    if(NSamples != NULL) return;    // already finalized!

// check if there is non zero number of bins per coordinate
    TotNBins = 1;
    for(int i=0; i < NCoords; i++) {
        TotNBins *= Sizes[i].NBins;
    }

// and now allocate data arrays
    NSamples.CreateVector(TotNBins);
}

//------------------------------------------------------------------------------

void CRSTAccumulator::Deallocate(void)
{
// do not destroy Sizes array !

// destroy only data arrays
    NSamples.FreeVector();
    TotNBins = 0;
}

//==============================================================================
//------------------------------------------------------------------------------
//==============================================================================

File: src/lib/fpmf/methods/rst/rst_accumulator.f90
!===============================================================================
! PMFLib - Library Supporting Potential of Mean Force Calculations
!-------------------------------------------------------------------------------
!    Copyright (C) 2011-2015 Petr Kulhanek, kulhanek@chemi.muni.cz
!    Copyright (C) 2013-2015 Letif Mones, lam81@cam.ac.uk
!    Copyright (C) 2007 Petr Kulhanek, kulhanek@enzim.hu
!    Copyright (C) 2006 Petr Kulhanek, kulhanek@chemi.muni.cz &
!                       Martin Petrek, petrek@chemi.muni.cz 
!    Copyright (C) 2005 Petr Kulhanek, kulhanek@chemi.muni.cz
!
!    This library is free software; you can redistribute it and/or
!    modify it under the terms of the GNU Lesser General Public
!    License as published by the Free Software Foundation; either
!    version 2.1 of the License, or (at your option) any later version.
!
!    This library is distributed in the hope that it will be useful,
!    but WITHOUT ANY WARRANTY; without even the implied warranty of
!    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
!    Lesser General Public License for more details.
!
!    You should have received a copy of the GNU Lesser General Public
!    License along with this library; if not, write to the Free Software
!    Foundation, Inc., 51 Franklin Street, Fifth Floor, 
!    Boston, MA  02110-1301  USA
!===============================================================================

module rst_accumulator

use pmf_constants

implicit none
contains

!===============================================================================
! Subroutine:  rst_accumulator_init
!===============================================================================

subroutine rst_accumulator_init()

    use rst_dat
    use pmf_dat
    use pmf_utils

    implicit none
    integer              :: i,tot_nbins
    integer              :: alloc_failed
    ! --------------------------------------------------------------------------

    ! only if we would like to update restart file and of if restart is explicitly required
    if( (fhistupdate .eq. 0) .and. (frestart .eqv. .false.) ) return

    ! init dimensions ------------------------------
    allocate(Accumulator%sizes(NumOfRSTItems), stat = alloc_failed)
    if( alloc_failed .ne. 0 ) then
        call pmf_utils_exit(PMF_OUT, 1,'[RST] Unable to allocate memory for rst accumulator!')
    endif

    tot_nbins = 1
    do i=1,NumOfRSTItems
        Accumulator%sizes(i)%min_value  = RSTCVList(i)%min_value
        Accumulator%sizes(i)%max_value  = RSTCVList(i)%max_value
        Accumulator%sizes(i)%nbins      = RSTCVList(i)%nbins
        Accumulator%sizes(i)%width      = abs(Accumulator%sizes(i)%max_value - Accumulator%sizes(i)%min_value)
        Accumulator%sizes(i)%bin_width  = Accumulator%sizes(i)%width / Accumulator%sizes(i)%nbins
        tot_nbins = tot_nbins * Accumulator%sizes(i)%nbins
    end do

    Accumulator%tot_nbins = tot_nbins

    ! RST force arrays
    allocate( Accumulator%nsamples(Accumulator%tot_nbins), &
              stat = alloc_failed)

    if( alloc_failed .ne. 0 ) then
        call pmf_utils_exit(PMF_OUT, 1,'[RST] Unable to allocate memory for rst accumulator (rstforce)!')
    endif

    call rst_accumulator_clear()

    return

end subroutine rst_accumulator_init

!===============================================================================
! Subroutine:  rst_accumulator_clear
!===============================================================================

subroutine rst_accumulator_clear()

    use rst_dat
    use pmf_dat

    implicit none
    ! --------------------------------------------------------------------------

    Accumulator%nsamples(:) = 0

end subroutine rst_accumulator_clear

!===============================================================================
! Function:  accumulator_index
! Arguments:
!               idxcoord ... number of ksi coordinate
!               accuvalue    ... value that is used to compute the bin index
! compute index for one accumulator coordinate
! Return value:     0,1,2, ..., sizes(idxcoord)%numbins-1
!===============================================================================

integer function rst_accumulator_index(idxcoord,accuvalue)

    use rst_dat
    use pmf_dat

    implicit none
    integer        :: idxcoord
    real(PMFDP)    :: accuvalue
    ! --------------------------------------------------------------------------

    ! we need number from zero - therefore we use floor(x)
    rst_accumulator_index = floor((accuvalue - Accumulator%sizes(idxcoord)%min_value) / &
                               Accumulator%sizes(idxcoord)%bin_width)

    if( rst_accumulator_index .lt. 0 .or. rst_accumulator_index .ge.  Accumulator%sizes(idxcoord)%nbins) then
        rst_accumulator_index = -1
        return
    end if

    ! do not try to include right boundary, since it will include the whole additional bin !

    return

end function rst_accumulator_index

!===============================================================================
! Function:  accumulator_globalindex
! Description:  Compute globalindex for accumulator, based on accuvalues of all coordinates
! Arguments:    none
! Return value: 1,2, ..., totalbins
!===============================================================================

integer function rst_accumulator_globalindex(lvalues)

    use rst_dat
    use pmf_dat

    implicit none
    real(PMFDP)            :: lvalues(:)
    ! -----------------------------------------------
    integer                :: idx_local,i
    ! --------------------------------------------------------------------------

    rst_accumulator_globalindex = 0

    do i=1,NumOfRSTItems
        idx_local = rst_accumulator_index(i,lvalues(i))

        if (idx_local .eq. -1) then
            rst_accumulator_globalindex = -1
            return
        end if

        rst_accumulator_globalindex = rst_accumulator_globalindex*Accumulator%sizes(i)%nbins + idx_local
    end do

    rst_accumulator_globalindex = rst_accumulator_globalindex + 1

    return

end function rst_accumulator_globalindex

!===============================================================================
! Subroutine:  rst_accumulator_read
!===============================================================================

subroutine rst_accumulator_read(iounit)

    use rst_dat
    use pmf_dat
    use pmf_utils

    implicit none
    integer             :: iounit
    ! -----------------------------------------------
    integer             :: ibitem
    character(len=4)    :: srst
    character(len=2)    :: sver
    character(len=80)   :: buffer
    ! --------------------------------------------------------------------------

    read(iounit,'(A80)') buffer

    ! read header - V1 --------------------------
    read(buffer,20,end=200,err=200) srst, sver, ibitem

    if( trim(adjustl(srst)) .ne. 'RST' ) then
        write(PMF_OUT,*) '[RST] header v1 key = [',srst,']'
        call pmf_utils_exit(PMF_OUT,1,'[RST] Missing RST key in RST accumulator v1 header!')
    end if

    if( trim(adjustl(sver)) .ne. 'V1' ) then
        write(PMF_OUT,*) '[RST] header v1 ver = [',sver,']'
        call pmf_utils_exit(PMF_OUT,1,'[RST] Unsupported version key in RST accumulator v1 header!')
    end if

    if( ibitem .ne. NumOfRSTItems ) then
        call pmf_utils_exit(PMF_OUT,1,'[RST] RST accumulator v1 contains different number of CVs!')
    end if

    call rst_accumulator_read_v1(iounit)
    return

    ! unsupported header - process error
200 continue
    call pmf_utils_exit(PMF_OUT,1,'[RST] Illegal header in RST accumulator (Only V1 header is supported)!')

20  format(A3,1X,A2,1X,I2)

end subroutine rst_accumulator_read

!===============================================================================
! Subroutine:  rst_accumulator_read_v1
!===============================================================================

subroutine rst_accumulator_read_v1(iounit)

    use rst_dat
    use pmf_dat
    use pmf_utils

    implicit none
    integer                         :: iounit
    ! -----------------------------------------------
    integer                         :: i,it,nbins
    character(len=PMF_MAX_TYPE)     :: stype
    character(len=PMF_MAX_CV_NAME)  :: sname
    real(PMFDP)                     :: min_value,max_value
    ! --------------------------------------------------------------------------

    ! read header --------------------------
    do i=1, NumOfRSTItems
        ! read CV definition
        read(iounit,20,end=102,err=102) it, stype, min_value, max_value, nbins
        ! check CV definition
        if( it .ne. i ) then
            call pmf_utils_exit(PMF_OUT,1,'[RST] Inccorect item in RST accumulator!')
        end if
        if( trim(adjustl(stype)) .ne. trim(RSTCVList(i)%cv%ctype) ) then
            write(PMF_OUT,*) '[RST] CV type = [',trim(adjustl(stype)),'] should be [',trim(RSTCVList(i)%cv%ctype),']'
            call pmf_utils_exit(PMF_OUT,1,'[RST] CV type was redefined in RST accumulator!')
        end if
        if( abs(min_value-Accumulator%sizes(i)%min_value) .gt. abs(Accumulator%sizes(i)%min_value/100000.0d0) ) then
            call pmf_utils_exit(PMF_OUT,1,'[RST] Minimal value of CV was redefined in RST accumulator!')
        end if
        if( abs(max_value-Accumulator%sizes(i)%max_value) .gt. abs(Accumulator%sizes(i)%max_value/100000.0d0) ) then
            call pmf_utils_exit(PMF_OUT,1,'[RST] Maximal value of CV was redefined in RST accumulator!')
        end if
        if( nbins .ne. Accumulator%sizes(i)%nbins ) then
            call pmf_utils_exit(PMF_OUT,1,'[RST] Number of CV bins was redefined in RST accumulator!')
        end if

        ! read names
        read(iounit,25,end=102,err=102) it, sname
        ! check names
        if( it .ne. i ) then
            call pmf_utils_exit(PMF_OUT,1,'[RST] Inccorect item in RST accumulator!')
        end if
        if( trim(adjustl(sname)) .ne. trim(RSTCVList(i)%cv%name) ) then
            write(PMF_OUT,*) '[RST] CV name = [',trim(adjustl(sname)),'] should be [',trim(RSTCVList(i)%cv%name),']'
            call pmf_utils_exit(PMF_OUT,1,'[RST] CV name was redefined in RST accumulator!')
        end if
    end do

    ! read accumulator data - RST forces
    if( Accumulator%tot_nbins .gt. 0 ) then
        read(iounit,30,end=100,err=100) (Accumulator%nsamples(i),i=1,Accumulator%tot_nbins)
    end if

    return

20  format(I2,1X,A10,1X,E18.11,1X,E18.11,1X,I6)
25  format(I2,1X,A55)
30  format(8(I9,1X))

100 call pmf_utils_exit(PMF_OUT,1,'[RST] Unable to read from RST accumulator v1 - data section!')
102 call pmf_utils_exit(PMF_OUT,1,'[RST] Unable to read from RST accumulator v1 - cv section!')

end subroutine rst_accumulator_read_v1

!===============================================================================
! Subroutine:  rst_accumulator_write
!===============================================================================

subroutine rst_accumulator_write(iounit)

    use rst_dat
    use pmf_dat
    use pmf_utils
    use pmf_unit

    implicit none
    integer                :: iounit
    ! -----------------------------------------------
    integer                :: i
    !---------------------------------------------------------------------------

    ! write header --------------------------
    write(iounit,10) 'RST', 'V1', NumOfRSTItems
    do i=1, NumOfRSTItems
        write(iounit,20) i,trim(RSTCVList(i)%cv%ctype), &
                          Accumulator%sizes(i)%min_value,Accumulator%sizes(i)%max_value, &
                          Accumulator%sizes(i)%nbins
        write(iounit,25) i,trim(RSTCVList(i)%cv%name)
    end do

    ! write accumulator data - RST samples
    write(iounit,30) (Accumulator%nsamples(i),i=1,Accumulator%tot_nbins)

    return

10  format(A3,1X,A2,1X,I2)
20  format(I2,1X,A10,1X,E18.11,1X,E18.11,1X,I6)
25  format(I2,1X,A55)
30  format(8(I9,1X))

end subroutine rst_accumulator_write

!===============================================================================
! Subroutine:  rst_accumulator_add_sample
!===============================================================================

subroutine rst_accumulator_add_sample(values)

    use rst_dat
    use pmf_dat

    implicit none
    real(PMFDP)    :: values(:)
    ! -----------------------------------------------
    integer        :: gi0
    ! --------------------------------------------------------------------------

    ! only if we would like to update restart file and of if restart is explicitly required
    if( (fhistupdate .eq. 0) .and. (frestart .eqv. .false.) ) return

    ! get global index to accumulator for average values within the set
    gi0 = rst_accumulator_globalindex(values)
    if( gi0 .le. 0 ) then
        outsidesamples = outsidesamples + 1
        return ! out of valid area
    else
        insidesamples = insidesamples + 1
    end if

    ! increase number of samples
    Accumulator%nsamples(gi0) = Accumulator%nsamples(gi0) + 1

end subroutine rst_accumulator_add_sample

!===============================================================================

end module rst_accumulator

